src/WebJobs.Extensions.DurableTask/Auth/AppAuthenticationCredentialsFactory.cs

```diff
@@ -0,0 +1,220 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Azure.Services.AppAuthentication;
+using Microsoft.Extensions.Logging;
+using Microsoft.WindowsAzure.Storage.Auth;
+
+namespace Microsoft.Azure.WebJobs.Extensions.DurableTask.Auth
+{
+    // TODO: Replace with Azure.Identity
+    internal sealed class AppAuthenticationCredentialsFactory : IStorageCredentialsFactory, IDisposable
+    {
+        private const string LoggerName = "Host.Triggers.DurableTask.Auth";
+
+        private readonly Func<string, AzureServiceTokenProvider> tokenProviderFactory;
+        private readonly ILogger logger;
+        private readonly AsyncLock cacheLock = new AsyncLock();
+        private readonly Dictionary<string, TokenCredential> cache = new Dictionary<string, TokenCredential>();
```

Comment:
The only reason I avoided concurrent dictionary was because I was wary of creating unnecessary access token requests when adding. If multiple asked for tokens and all attempted to update the cache, multiple threads may attempt to fetch the token but only 1 will persist in a cache. Hence why I locked on cache miss so that only 1 thread would ultimately make the access token request at a time.

Reply:
Looking at the extension library, I am unsure if this code needs to even be thread-safe. Could the connections be created by separate threads simultaneously?

Reply:
If you're asking about connections to Azure Storage, yes, those can be created by separate threads simultaneously. The problem with the locking logic you have below is that two threads can still race with each other: one calling `this.cache.TryGetValue` (outside the lock) and one calling `this.cache.Add` (from inside the lock). Concurrent Get/Add requests on a regular dictionary can cause problems such as infinite loops that eat up all the CPU (I've seen this in production systems before). For that reason, I suggest using `ConcurrentDictionary`.

Reply:
Ah, I wasn't aware of these issues, but it makes sense. I'll swap out of the dictionary implementation

Reply:
I no longer have a cache in my refactor. Instead I simply rely on the `AzureComponentFactory` to generate the connections, which seems to be consistent (I think) with the WebJobs SDK.